### PR TITLE
Adding image onFocus and allowing image container style props for DynamicCollage

### PR DESCRIPTION
**Pull request by harbolaez**
Adding `onImageFocus` and `imageFocussedStyle` prop to DynamicCollage. When a user pressed on a image it will trigger the onImafeFocus property exposing two parameters metrixId(m), imageIndex(i) and the event(e) of the TouchableWithoutFeedback. You can add custom style to the focussed image with the prop `imageFocussedStyle`

## API

```
<DynamicCollage
  images={[]}
  onImageFocus={({ e, m, i }) => {
    console.log(m,i)
  }}
  imageFocussedStyle={{ borderColor: "red" }}
  ...
/>
```

### CODE REVIEW

**Comment by harbolaez** [comment]
@lukebrandonfarrell let me know what you think? But looking at this and thinking about it again, I think it makes sense to have a `imageOnPress` callback too which lets the user set the style of that corresponding image or anything they want to do. Also maybe exposing the location of where the press was trigger might be worth it too. let me know what you think?

**Comment by lukebrandonfarrell** [comment]
Thanks for the PR @harbolaez ! Will test these changes this evening, then will merge and push if I discover no issues :)

File: src/CollageImage.js [hunk 1]
```diff
@@ -71,9 +71,10 @@ class CollageImage extends React.Component {
 
     this._panResponder = PanResponder.create({
       onStartShouldSetPanResponder: (evt, gestureState) => true,
-      onStartShouldSetPanResponderCapture: (evt, gestureState) => true,
       onMoveShouldSetPanResponder: (evt, gestureState) => true,
-      onMoveShouldSetPanResponderCapture: (evt, gestureState) => true,
+      onMoveShouldSetPanResponderCapture: (evt, gestureState) => {
+        return Math.abs(gestureState.dx) > 5;
+      },
```
Comment (lukebrandonfarrell) [inline src/CollageImage.js:77]: I propose we keep this as it was and handle the tap in `onPanResponderEnd`

```suggestion
      onMoveShouldSetPanResponderCapture: (evt, gestureState) => true,
```

File: src/CollageImage.js [hunk 2]
```diff
@@ -566,16 +579,18 @@ class CollageImage extends React.Component {
...
-            ref={"image"}
-            source={source}
-            style={[
-              style,
-              { right, bottom, width, height },
-              selected ? imageSelectedStyle : null,
-            ]}
-            resizeMode="cover"
-          />
+          <TouchableWithoutFeedback onPress={this.props.onImageFocus}>
```
Comment (lukebrandonfarrell) [inline src/CollageImage.js:582]: Not needed anymore if we add the handling in `onPanResponderEnd`
```suggestion
```

File: src/CollageImage.js [hunk 3]
```diff
@@ -566,16 +579,18 @@ class CollageImage extends React.Component {
...
+              ref={"image"}
+              source={source}
+              style={[
+                style,
+                { right, bottom, width, height },
+                selected ? imageSelectedStyle : null,
+              ]}
+              resizeMode="cover"
+            />
+          </TouchableWithoutFeedback>
```
Comment (lukebrandonfarrell) [inline src/CollageImage.js:593]:
```suggestion
```

File: src/DynamicCollage.js [hunk 4]
```diff
@@ -339,6 +362,8 @@ DynamicCollage.propTypes = {
   retainScaleOnSwap: PropTypes.bool,
   longPressDelay: PropTypes.number,
   longPressSensitivity: PropTypes.number, // 1 - 20 - How sensitive is the long press?
+  onImageFocus: PropTypes.func,
+  onImagePress: PropTypes.func,
```
Comment (lukebrandonfarrell) [inline src/DynamicCollage.js:366]: Is there any need for this? Can't find it being used
```suggestion
```

File: src/DynamicCollage.js [hunk 5]
```diff
@@ -339,6 +362,8 @@ DynamicCollage.propTypes = {
   retainScaleOnSwap: PropTypes.bool,
   longPressDelay: PropTypes.number,
   longPressSensitivity: PropTypes.number, // 1 - 20 - How sensitive is the long press?
+  onImageFocus: PropTypes.func,
```
Comment (lukebrandonfarrell) [inline src/DynamicCollage.js:365]: Could we also add `imageContainerStyle` and `imageFocussedStyle` here to the PropTypes?

And also if possible update the documentation README.md with their usage?

File: src/DynamicCollage.js [hunk 6]
```diff
@@ -305,6 +320,11 @@ DynamicCollage.defaultProps = {
     backgroundColor: "white",
   },
   imageStyle: {}, // DEFAULT IMAGE STYLE
+  imageContainerStyle: {
+    // DEFAULT IMAGE CONTAINER STYLE
+    borderWidth: 2,
+    borderColor: "white",
+  },
```
Comment (lukebrandonfarrell) [inline src/DynamicCollage.js:327]: What was the reason behind this? Can't seem to figure it out

File: src/DynamicCollage.js [hunk 7]
```diff
@@ -284,6 +289,16 @@ class DynamicCollage extends React.Component {
...
+
+  setImageFocusId(m, i) {
+    return function (e) {
+      const imageId = `image${m}-${i}`;
+      this.setState((prevState) => ({
+        imageFocusId: prevState.imageFocusId === imageId ? null : imageId,
+      }));
+      this.props.onImageFocus && this.props.onImageFocus({ e, m, i });
+    }.bind(this);
+  }
```
Comment (lukebrandonfarrell) [inline src/DynamicCollage.js:301]: A few things for this:

- Could we add a JSDoc comment explaining something about what this function is used for and the param types (good for IDE intelligence)
- What is the reason for wrapping this in a `function` with bind and returning it? Instead could we do:

```
 onImageFocus={() => this.setImageFocusId(m, i)}
...
setImageFocusId(m, i) {
    const imageId = `image${m}-${i}`;
    this.setState((prevState) => ({
      imageFocusId: prevState.imageFocusId === imageId ? null : imageId,
    }));

    this.props.onImageFocus && this.props.onImageFocus({ m, i });
}
```
The above is a bit easier to read and follow

**Review by lukebrandonfarrell — CHANGES_REQUESTED** [review]
Nice work on this PR! This functionality was actually quite complex as you had to track the selected ID of the image. Here are my review comments:

I propose that instead of wrapping with a touchable we handle the click inside `onPanResponderEnd` i.e. like below (I tested this):

```js
 onPanResponderEnd: (e, gestureState) => {
       // Focusses the image, if user just taps it, and there is no movement or holding
        if(Math.abs(gestureState.dx) < 5 && !this.state.selected) {
          if(this.props.onImageFocus) this.props.onImageFocus();
        }
         // Disable scaling, and panning
        this.panning = false;
        this.scaling = false;
       ...
```

File: src/DynamicCollage.js (hunk 6 above)
Comment (harbolaez) [reply src/DynamicCollage.js:327]: Yeah - this was to allow end-user to add their custom border to the image container while keeping the default as you had before.

File: src/DynamicCollage.js (hunk 7 above)
Comment (harbolaez) [reply src/DynamicCollage.js:301]: Yeah that removing the touchable would do it. I wanted to expose the `event` on where the press happens, just incase we wanted to show a tooltip, etc. But that' fine for now.

**Comment by lukebrandonfarrell** [comment]
@harbolaez Will review this today :)

File: src/DynamicCollage.js (hunk 7 above)
Comment (lukebrandonfarrell) [reply src/DynamicCollage.js]: Makes sense! Do you think we can update this block to this? It's more readable and inline with the other callback functions.

```
onImageFocus={(event) => this.setImageFocusId(event, m, i)}

...

setImageFocusId(event, m, i) {
    const imageId = `image${m}-${i}`;
    this.setState((prevState) => ({
      imageFocusId: prevState.imageFocusId === imageId ? null : imageId,
    }));

    this.props.onImageFocus && this.props.onImageFocus(event, m, i);
}
```

File: src/DynamicCollage.js (hunk 6 above)
Comment (lukebrandonfarrell) [reply src/DynamicCollage.js:342]: Sounds good!

File: src/DynamicCollage.js (hunk 7 above)
Comment (lukebrandonfarrell) [reply src/DynamicCollage.js]: After that, this looks good to me!

**Comment by lukebrandonfarrell** [comment]
@harbolaez if you could also resolve the conflicts, that would be great! Then I'll create a new version on NPM for all three of these features #21 #22 and #20

**Comment by lukebrandonfarrell** [comment]
@all-contributors please add @harbolaez for code and documentation

File: src/DynamicCollage.js [hunk 8]
```diff
@@ -2,13 +2,15 @@ import React from "react";
...
 
+import { LayoutData } from "./Layouts";
 import CollageImage from "./CollageImage";
 
 class DynamicCollage extends React.Component {
   constructor(props) {
     super(props);
 
     this.state = {
+      imageFocusId: "",
```
Comment (lukebrandonfarrell) [inline src/DynamicCollage.js:13]:
```suggestion
      imageFocusId: null,
```

File: README.md [hunk 9]
```diff
@@ -134,6 +134,8 @@ collageRef.current.replaceImage("https://picsum.photos/200", m, i);
 | imageSwapStyleReset  | object   | Yes      | style    | The reverse of imageSwapStyle to reset style after swap. Vital for direct manipulation.                                   |
 | separatorStyle       | object   | Yes      | style    | Style applied to image container. Use border width to create margin between images.                                       |
 | containerStyle       | object   | Yes      | style    | Style applied to the container of the collage. Collage border can be applied here.                                        |
+| imageContainerStyle  | object   | Yes      | style    | Style applied to the image container of the collage.                                                                      |
```
Comment (lukebrandonfarrell) [inline README.md:137]:
```suggestion
| imageContainerStyle  | object   | Yes      | style    | Style applied to each image container.                                                                      |
```

File: README.md [hunk 10]
```diff
@@ -134,6 +134,8 @@ collageRef.current.replaceImage("https://picsum.photos/200", m, i);
...
+| imageFocussedStyle   | object   | Yes      | style    | Style applied to the focused image container of the collage.                                                              |
```
Comment (lukebrandonfarrell) [inline README.md:138]:
```suggestion
| imageFocussedStyle   | object   | Yes      | style    | Style applied to the focused image container.                                                              |
```

**Review by lukebrandonfarrell — APPROVED** [review]
Looks good to me! 💯 

Just need to run a local test then will publish this! 👍

**Pull request merged**